### DISCUSSION TITLE
Question about multiple editors.

### DISCUSSION THREAD

**Discussion by Jxhnn** [mkslanc/ace-linters#101] [Q&A] [answered]
Hello, i have a question about this awesome project.

Can the linter support multiple file modules ? 
Like i want to put in my editor a typescript file with an import. Can it be also parsed ? Is it possible to add files like that without creating an editor ? 

Best regards.

**Comment by mkslanc** [comment, marked as answer]
@Jxhnn, thank you for your interest in this project! Due to limitations in the browser environment, achieving a fully functional experience, especially with the file system, is somewhat challenging. However, there is a workaround for the TypeScript service: you can set your files to the `extraLibs` property in `globalOptions`, similar to this:
```typescript
languageProvider.setGlobalOptions("typescript", {
    extraLibs: {
        "dir/someFile.ts": {
            content: "export var someVar = 123",
            version: 1
        }
    }
});
```

import in that case could be:
```typescript
import * as someFile from "./dir/someFile";

someFile.someVar
```

**Reply by josheyr** [reply]
Is anything similar possible with javascript? Looking to add globals, I saw a "globals" configuration in ESlint config but can't get it to work.

**Reply by josheyr** [reply]
this is what I tried:
```javascript
provider.setGlobalOptions("javascript", {
            globals: {
                   "var1": "readonly"
            }
        })
```

**Reply by mkslanc** [reply]
I am not sure it would be possible to achieve same level of checks and integration with eslint, but you could easily use `typescript` service for any javascript files by setting `allowJs` to `true` ([Example](https://mkslanc.github.io/ace-playground/?value=Ly9zYW1wbGVzL3R5cGVzY3JpcHQgbGFuZ3VhZ2Ugc2VydmljZQoKYWNlLnJlcXVpcmUoImFjZS9leHQvbGFuZ3VhZ2VfdG9vbHMiKTsKbGV0IGVkaXRvciA9IGFjZS5lZGl0KCJleGFtcGxlIiwgewogICAgdGhlbWU6ICJhY2UvdGhlbWUvdGV4dG1hdGUiLAogICAgbW9kZTogImFjZS9tb2RlL2phdmFzY3JpcHQiLAogICAgdmFsdWU6IGBpbXBvcnQgKiBhcyBsaWJGaWxlIGZyb20gIi4vZGlyL2ZpbGUiOwoKbGV0IHRlc3Q7CmxldCBjaGFpbmFibGUgPSBuZXcgQ2hhaW5hYmxlT25lKCk7CmNoYWluYWJsZS5jaGFpbmFibGVUd28uYWRkQWxwaGEoInRlc3QiKS5zZXRCZXRhKCk7CmxpYkZpbGUuZGF0YS5zZXRBbHBoYSgiIikuY2hhaW5hYmxlVHdvO2AKfSk7Ci8vIGVuYWJsZSBhdXRvY29tcGxldGlvbiBhbmQgc25pcHBldHMKZWRpdG9yLnNldE9wdGlvbnMoewogICAgZW5hYmxlQmFzaWNBdXRvY29tcGxldGlvbjogdHJ1ZSwKICAgIGVuYWJsZVNuaXBwZXRzOiB0cnVlLAogICAgZW5hYmxlTGl2ZUF1dG9jb21wbGV0aW9uOiB0cnVlCn0pOwoKLy9jcmVhdGUgTGFuZ3VhZ2UgUHJvdmlkZXIgZm9yIGVkaXRvcgpsZXQgcHJvdmlkZXIgPSBMYW5ndWFnZVByb3ZpZGVyLmZyb21DZG4oImh0dHBzOi8vY2RuLmpzZGVsaXZyLm5ldC9ucG0vYWNlLWxpbnRlcnMvYnVpbGQiKTsKCi8vIFJlZ2lzdGVyIHRoZSBlZGl0b3Igd2l0aCB0aGUgbGFuZ3VhZ2UgcHJvdmlkZXIKcHJvdmlkZXIucmVnaXN0ZXJFZGl0b3IoZWRpdG9yKTsKCnByb3ZpZGVyLnNldEdsb2JhbE9wdGlvbnMoInR5cGVzY3JpcHQiLCB7CiAgICBlcnJvckNvZGVzVG9UcmVhdEFzV2FybmluZzogWwogICAgICAgICI2MTMzIgogICAgXSwKICAgIGNvbXBpbGVyT3B0aW9uczogewogICAgICAgIG5vVW51c2VkTG9jYWxzOiB0cnVlLAogICAgICAgIGFsbG93SnM6IHRydWUKICAgIH0sCiAgICBleHRyYUxpYnM6IHsKICAgICAgICAibGliRGVjbGFyYXRpb24uZC50cyI6IHsKICAgICAgICAgICAgY29udGVudDoKYGRlY2xhcmUgY2xhc3MgQ2hhaW5hYmxlT25lIHsKICAgIGNoYWluYWJsZVR3bzogQ2hhaW5hYmxlVHdvOwogICAgc2V0QWxwaGEodmFsdWU6IHN0cmluZyk6IHRoaXM7CiAgICBzZXRCZXRhKHZhbHVlOiBudW1iZXIpOiBDaGFpbmFibGVUd287Cn0KCmRlY2xhcmUgY2xhc3MgQ2hhaW5hYmxlVHdvIHsKICAgIHNldEdhbW1hKHZhbHVlOiBib29sZWFuKTogdGhpczsKICAgIGFkZEFscGhhKHZhbHVlOiBzdHJpbmcpOiBDaGFpbmFibGVPbmU7Cn1gLAogICAgICAgICAgICB2ZXJzaW9uOiAxCiAgICAgICAgfSwKICAgICAgICAiZGlyL2ZpbGUuanMiOiB7CiAgICAgICAgICAgIGNvbnRlbnQ6ICJleHBvcnQgdmFyIGRhdGEgPSBuZXcgQ2hhaW5hYmxlT25lKCk7IiwKICAgICAgICAgICAgdmVyc2lvbjogMQogICAgICAgIH0KICAgIH0KfSk7ClwwXDA8c2NyaXB0IHNyYz0iaHR0cHM6Ly93d3cudW5wa2cuY29tL2FjZS1idWlsZHNAbGF0ZXN0L3NyYy1ub2NvbmZsaWN0L2FjZS5qcyI%2BPC9zY3JpcHQ%2BCjwhLS0gbG9hZCBhY2UgbGFuZ3VhZ2UgdG9vbHMgLS0%2BCjxzY3JpcHQgc3JjPSJodHRwczovL3d3dy51bnBrZy5jb20vYWNlLWJ1aWxkc0BsYXRlc3Qvc3JjLW5vY29uZmxpY3QvZXh0LWxhbmd1YWdlX3Rvb2xzLmpzIj48L3NjcmlwdD4KPCEtLSBsb2FkIGFjZSBsaW50ZXJzIC0tPgo8c2NyaXB0IHNyYz0iaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9hY2UtbGludGVycyI%2BPC9zY3JpcHQ%2BCjxkaXYgaWQ9J2V4YW1wbGUnIHN0eWxlPSJoZWlnaHQ6IDEwMCUiPjwvZGl2Pgo%3D#typescript-language-service)).

**Reply by josheyr** [reply]
Thanks I'll give it a try. Do you by any chance know if it's possible for one to fork the library and edit the eslint configuration? Such as environment etc. https://eslint.org/docs/latest/use/configure/language-options
(using npm run build)

**Reply by mkslanc** [reply]
[Better example](https://mkslanc.github.io/ace-playground/?value=Ly9zYW1wbGVzL3R5cGVzY3JpcHQgbGFuZ3VhZ2Ugc2VydmljZQoKYWNlLnJlcXVpcmUoImFjZS9leHQvbGFuZ3VhZ2VfdG9vbHMiKTsKbGV0IGVkaXRvciA9IGFjZS5lZGl0KCJleGFtcGxlIiwgewogICAgdGhlbWU6ICJhY2UvdGhlbWUvdGV4dG1hdGUiLAogICAgbW9kZTogImFjZS9tb2RlL2phdmFzY3JpcHQiLAogICAgdmFsdWU6IGB2YXIgbGliRmlsZSA9IHJlcXVpcmUoIi4vZGlyL2ZpbGUuanMiKTsKCmNvbnNvbGUubG9nKGxpYkZpbGUuZGF0YSk7YAp9KTsKLy8gZW5hYmxlIGF1dG9jb21wbGV0aW9uIGFuZCBzbmlwcGV0cwplZGl0b3Iuc2V0T3B0aW9ucyh7CiAgICBlbmFibGVCYXNpY0F1dG9jb21wbGV0aW9uOiB0cnVlLAogICAgZW5hYmxlU25pcHBldHM6IHRydWUsCiAgICBlbmFibGVMaXZlQXV0b2NvbXBsZXRpb246IHRydWUKfSk7CgovL2NyZWF0ZSBMYW5ndWFnZSBQcm92aWRlciBmb3IgZWRpdG9yCmxldCBwcm92aWRlciA9IExhbmd1YWdlUHJvdmlkZXIuZnJvbUNkbigiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9hY2UtbGludGVycy9idWlsZCIpOwoKLy8gUmVnaXN0ZXIgdGhlIGVkaXRvciB3aXRoIHRoZSBsYW5ndWFnZSBwcm92aWRlcgpwcm92aWRlci5yZWdpc3RlckVkaXRvcihlZGl0b3IpOwoKcHJvdmlkZXIuc2V0R2xvYmFsT3B0aW9ucygidHlwZXNjcmlwdCIsIHsKICAgIGVycm9yQ29kZXNUb1RyZWF0QXNXYXJuaW5nOiBbCiAgICAgICAgIjYxMzMiCiAgICBdLAogICAgY29tcGlsZXJPcHRpb25zOiB7CiAgICAgICAgbm9VbnVzZWRMb2NhbHM6IHRydWUsCiAgICAgICAgYWxsb3dKczogdHJ1ZSwKICAgIH0sCiAgICBleHRyYUxpYnM6IHsKICAgICAgICAiZGlyL2ZpbGUuanMiOiB7CiAgICAgICAgICAgIGNvbnRlbnQ6ICJleHBvcnRzLmRhdGEgPSAndGVzdCciLAogICAgICAgICAgICB2ZXJzaW9uOiAxCiAgICAgICAgfQogICAgfQp9KTsKXDBcMDxzY3JpcHQgc3JjPSJodHRwczovL3d3dy51bnBrZy5jb20vYWNlLWJ1aWxkc0BsYXRlc3Qvc3JjLW5vY29uZmxpY3QvYWNlLmpzIj48L3NjcmlwdD4KPCEtLSBsb2FkIGFjZSBsYW5ndWFnZSB0b29scyAtLT4KPHNjcmlwdCBzcmM9Imh0dHBzOi8vd3d3LnVucGtnLmNvbS9hY2UtYnVpbGRzQGxhdGVzdC9zcmMtbm9jb25mbGljdC9leHQtbGFuZ3VhZ2VfdG9vbHMuanMiPjwvc2NyaXB0Pgo8IS0tIGxvYWQgYWNlIGxpbnRlcnMgLS0%2BCjxzY3JpcHQgc3JjPSJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL2FjZS1saW50ZXJzIj48L3NjcmlwdD4KPGRpdiBpZD0nZXhhbXBsZScgc3R5bGU9ImhlaWdodDogMTAwJSI%2BPC9kaXY%2BCg%3D%3D#typescript-language-service)

**Reply by mkslanc** [reply]
If you need to change only configs, it could be easily achieved by using - `setGlobalOptions`, for eslint it would take [such options](https://github.com/mkslanc/ace-linters/blob/3b4ef527e88cad211b8f65401b5a34f649d8f4b3/packages/ace-linters/src/types/language-service.ts#L139). But, if you want to use different version or more global changes, you could build your own linter similar to [this one](https://github.com/mkslanc/ace-linters/blob/a68523ef21747dc0865588f47c47667a4ffc60b3/packages/ace-sql-linter/README.md#L1). [Bundled browserified eslint](https://github.com/mkslanc/ace-linters/tree/7ae90fa0b3d55feb0d12640d2947a092c75aa1a1/packages/eslint-linter-bundle)

**Reply by josheyr** [reply]
Thanks for your quick support.

I'm trying to use setGlobalOptions since one of these is "globals" which I'm guessing should work like this:
```
        let provider = LanguageProvider.fromCdn("https://www.unpkg.com/ace-linters@latest/build/");

        let editor = ace.edit(this);
        editor.setOptions({
            maxLines: 1000,
            minLines: 10,
            autoScrollEditorIntoView: true,
            enableBasicAutocompletion: true,
            mode: "ace/mode/javascript",
            enableSnippets: true,
            enableLiveAutocompletion: true,
        });

        provider.registerEditor(editor);

        provider.setGlobalOptions("javascript", {
            globals: {
                "globalTestVar": "readonly",
            }
        })
```

I'm just still getting "Cannot find name" from the linter. Am I using it correctly?
![image](https://github.com/mkslanc/ace-linters/assets/54540136/44a917af-6ddf-47fc-b14f-a55b91bf9384)

**Reply by mkslanc** [reply]
Ok, I got what the problem here -> by default typescript service is more prior than eslint one. So actually you are using typescript in your test, but configuring eslint. To use only eslint you could do something like [this](https://mkslanc.github.io/ace-playground/?value=Ly9zYW1wbGVzL3R5cGVzY3JpcHQgbGFuZ3VhZ2Ugc2VydmljZQoKYWNlLnJlcXVpcmUoImFjZS9leHQvbGFuZ3VhZ2VfdG9vbHMiKTsKbGV0IGVkaXRvciA9IGFjZS5lZGl0KCJleGFtcGxlIiwgewogICAgdGhlbWU6ICJhY2UvdGhlbWUvdGV4dG1hdGUiLAogICAgbW9kZTogImFjZS9tb2RlL2phdmFzY3JpcHQiLAogICAgdmFsdWU6IGB2YXIgbGliRmlsZSA9IHJlcXVpcmUoIi4vZGlyL2ZpbGUuanMiKTsKCmNvbnNvbGUubG9nKGxpYkZpbGUuZGF0YSk7CgpnbG9iYWxUZXN0VmFyCmdsb2JhbFRlc3RCYXJgCn0pOwovLyBlbmFibGUgYXV0b2NvbXBsZXRpb24gYW5kIHNuaXBwZXRzCmVkaXRvci5zZXRPcHRpb25zKHsKICAgIGVuYWJsZUJhc2ljQXV0b2NvbXBsZXRpb246IHRydWUsCiAgICBlbmFibGVTbmlwcGV0czogdHJ1ZSwKICAgIGVuYWJsZUxpdmVBdXRvY29tcGxldGlvbjogdHJ1ZQp9KTsKCi8vY3JlYXRlIExhbmd1YWdlIFByb3ZpZGVyIGZvciBlZGl0b3IKbGV0IHByb3ZpZGVyID0gTGFuZ3VhZ2VQcm92aWRlci5mcm9tQ2RuKHsgc2VydmljZXM6IFtdLCAKc2VydmljZU1hbmFnZXJDZG46ICJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL2FjZS1saW50ZXJzL2J1aWxkIiwgCmluY2x1ZGVEZWZhdWx0TGludGVyczogeyBqYXZhc2NyaXB0OiB0cnVlIH0gfSk7CgovLyBSZWdpc3RlciB0aGUgZWRpdG9yIHdpdGggdGhlIGxhbmd1YWdlIHByb3ZpZGVyCnByb3ZpZGVyLnJlZ2lzdGVyRWRpdG9yKGVkaXRvcik7Cgpwcm92aWRlci5zZXRHbG9iYWxPcHRpb25zKCJqYXZhc2NyaXB0IiwgewogICAgZ2xvYmFsczogewogICAgICAgICJnbG9iYWxUZXN0VmFyIjogInJlYWRvbmx5IiwKICAgIH0KfSlcMFwwPHNjcmlwdCBzcmM9Imh0dHBzOi8vd3d3LnVucGtnLmNvbS9hY2UtYnVpbGRzQGxhdGVzdC9zcmMtbm9jb25mbGljdC9hY2UuanMiPjwvc2NyaXB0Pgo8IS0tIGxvYWQgYWNlIGxhbmd1YWdlIHRvb2xzIC0tPgo8c2NyaXB0IHNyYz0iaHR0cHM6Ly93d3cudW5wa2cuY29tL2FjZS1idWlsZHNAbGF0ZXN0L3NyYy1ub2NvbmZsaWN0L2V4dC1sYW5ndWFnZV90b29scy5qcyI%2BPC9zY3JpcHQ%2BCjwhLS0gbG9hZCBhY2UgbGludGVycyAtLT4KPHNjcmlwdCBzcmM9Imh0dHBzOi8vY2RuLmpzZGVsaXZyLm5ldC9ucG0vYWNlLWxpbnRlcnMiPjwvc2NyaXB0Pgo8ZGl2IGlkPSdleGFtcGxlJyBzdHlsZT0iaGVpZ2h0OiAxMDAlIj48L2Rpdj4K#typescript-language-service). I will correct wrong function overload in next version.

**Reply by josheyr** [reply]
Perfect, thank you so much!

**Reply by josheyr** [reply]
While it's no problem for my use case, I think it lost the fancy error hover tooltips which I thought was a nice touch. Is this available with just eslint?
![image](https://github.com/mkslanc/ace-linters/assets/54540136/e623294b-1aa1-45b4-b999-f249fb1e798a)

**Reply by josheyr** [reply]
Also hasn't got autocomplete anymore with my settings. If none is available with eslint is there a way I can make it only use my editor.completers rather than the linter autocompletion?

**Reply by mkslanc** [reply]
EsLint provides [only diagnostics](https://github.com/mkslanc/ace-linters/blob/32edf72ffadeef3dad5eeedc65dfa22cf61fb03b/packages/ace-linters/src/services/javascript/javascript-service.ts#L64), you could use either `typescript` service, or `eslint` in conjunction with `typescript` - it would use diagnostics from eslint and other functionality from typescript

**Reply by josheyr** [reply]
Can I disable the override of no autocomplete and use the default ace editor mode?

**Reply by mkslanc** [reply]
Sure - just set `completion` and `completionResolve` to `false`. 
```javascript
let provider = LanguageProvider.fromCdn({ services: [], 
serviceManagerCdn: "https://cdn.jsdelivr.net/npm/ace-linters/build", 
includeDefaultLinters: { javascript: true } }, {functionality: {
    completion: false, completionResolve: false
}});
```

**Reply by josheyr** [reply]
Worked like a charm, really appreciate the quick and helpful response. Cheers!

**Reply by Jxhnn** [reply]
@mkslanc Hello, sorry for the inconvenience but I don't think this is still working. I can onlym make working imports with absolute paths and not relative ones. Is it possible to check this please ? Thank you :)

**Reply by mkslanc** [reply]
@Jxhnn there was small bug, hopefully should be resolved in [1.3.2](https://github.com/mkslanc/ace-linters/releases/tag/v1.3.2) . Also if you are using multiple editors we have new API endpoint to [set file path associated with current session](https://github.com/mkslanc/ace-linters/blob/main/packages/ace-linters/types/language-provider.d.ts#L54). I updated default demo with possible use cases for [extralibs](https://github.com/mkslanc/ace-linters/blob/main/packages/demo/webworker-lsp/demo.ts#L57) and [`setSessionFilePath` itself](https://github.com/mkslanc/ace-linters/blob/main/packages/demo/utils.ts#L61)

**Reply by mkslanc** [reply]
![image](https://github.com/user-attachments/assets/e77d62b0-1790-4313-813a-defe81305a22)

**Reply by Jxhnn** [reply]
Thank you very much !!
